### PR TITLE
Descriptor overhaul

### PR DESCRIPTION
**Pull request by GambolingPangolin**
This change fixes the definition of descriptors so that every `OutputDescriptor` value corresponds to a valid output (or family).  Also, I added some utility functions:

* Get an address from an output descriptor
* Get a script from a script descriptor
* Check if a key descriptor is definite (not a family)
* Make a key descriptor definite

I also changed the miniscript compiler to use `OP_1` ... `OP_16` if possible instead of pushing a `CScriptNum`.

### CODE REVIEW

File: bitcoin-scripting.cabal [hunk 1]
```diff
@@ -12,19 +12,17 @@ category:            Language
 build-type:          Simple
 extra-source-files:  CHANGELOG.md
 
-tested-with: GHC == 8.8.4
+tested-with: GHC == 8.10.4
 
 common base
   default-language: Haskell2010
-  ghc-options:      -Wall
+  ghc-options: -Wall -Wunused-packages -Werror
```
Comment (wraithm) [inline bitcoin-scripting.cabal]: There's a bunch of other stuff you could add if you wanted.

```
    - -Wall
    - -Werror
    - -Wmissing-home-modules
    - -Widentities
    - -Wincomplete-uni-patterns
    - -Wincomplete-record-updates
    - -Wpartial-fields
    - -Wmissing-export-lists
    - -fwarn-tabs
    - -fno-warn-unused-do-bind
    - -funbox-strict-fields
    - -Wunused-packages
 ```

that might be a good list

File: src/Language/Bitcoin/Script/Utils.hs [hunk 2]
```diff
@@ -33,7 +33,9 @@ toCScriptNum n
     b = BS.snoc b' msb
 
 pushNumber :: Int -> ScriptOp
-pushNumber = opPushData . toCScriptNum
+pushNumber i
+    | i <= 16 = intToScriptOp i
```
Comment (wraithm) [inline src/Language/Bitcoin/Script/Utils.hs:39]: This seems very curious to me. Would you never `OP_PUSHDATA` on a number `<= 16`? Maybe I just don't know.

Comment (wraithm) [reply src/Language/Bitcoin/Script/Utils.hs:39]: Mixing op codes in this function seems strange is really what I'm saying.

Comment (GambolingPangolin) [reply src/Language/Bitcoin/Script/Utils.hs:39]: The goal here is to push bytes to the stack that a later opcode will interpret as a number.  The `OP_i` opcodes take one byte to push their number, versus the PUSHDATA opcodes, which require at least 2 bytes even for small numbers.

File: bitcoin-scripting.cabal (hunk 1 above)
Comment (GambolingPangolin) [reply bitcoin-scripting.cabal]: Excellent.  I'll add them.

Comment (GambolingPangolin) [reply bitcoin-scripting.cabal]: Some of these are redundant: https://downloads.haskell.org/ghc/8.10.4/docs/html/users_guide/using-warnings.html#options-sanity

Comment (wraithm) [reply bitcoin-scripting.cabal]: (turned out the only one that was redundant was warn tabs!)

**Review by wraithm — APPROVED** [review]

**Pull request merged**